File: golang/graph_general/clone_graph.go
package main

type Node struct {
	Val       int
	Neighbors []*Node
}

func cloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	cloned := make(map[*Node]*Node)

	var dfs func(*Node) *Node

	dfs = func(n *Node) *Node {
		if n == nil {
			return nil
		}

		if clone, found := cloned[n]; found {
			return clone
		}

		copy := &Node{Val: n.Val}

		cloned[n] = copy

		for _, neighbor := range n.Neighbors {
			copy.Neighbors = append(copy.Neighbors, dfs(neighbor))
		}

		return copy
	}

	return dfs(node)
}

File: python/graph_general/clone_graph.py
from typing import List, Optional

class Node:
    def __init__(self, val: int = 0, neighbors: Optional[List['Node']] = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self,node: 'Node') -> 'Node':
        if not node:
            return None

        cloned = {}


        def dfs(n: 'Node') -> 'Node':
            if n in cloned:
                return cloned[n]

            copy = Node(n.val)

            cloned[n] = copy

            for neighbor in n.neighbors:
                copy.neighbors.append(dfs(neighbor))


            return copy

        return dfs(node)
